File: feature/vault/src/main/kotlin/xyz/libravault/feature/vault/VaultReaderViewModel.kt
package xyz.libravault.feature.vault

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.readium.r2.shared.publication.Publication
import xyz.libravault.core.ui.theme.ReadingTheme
import xyz.libravault.core.vaultcrypto.VaultFileReader
import xyz.libravault.core.vaultstore.VaultBookmark
import xyz.libravault.core.vaultstore.VaultHighlight
import xyz.libravault.core.vaultstore.VaultSessionManager
import xyz.libravault.core.vaultstore.VaultStore
import javax.inject.Inject

sealed class VaultReaderState {
    object Loading : VaultReaderState()
    data class Error(val message: String) : VaultReaderState()
    /** Publication is DRM-restricted (e.g. Adobe ADEPT, LCP) — Libravault has no
     * decryption support, so it's surfaced distinctly from [Error] to show
     * DRM-specific copy instead of a raw parser error message. */
    data class DrmProtected(val schemeName: String?) : VaultReaderState()
    data class EpubReady(val title: String, val publication: Publication) : VaultReaderState()
    /** [reader] stays open for as long as this state is current — the PDF
     * screen reads from it directly (see [VaultReaderViewModel.pdfReader]),
     * unlike EPUB where Readium owns the read lifecycle once [Publication]
     * exists. */
    data class PdfReady(val title: String) : VaultReaderState()
    /** Not a reading format — the caller should have routed audio to
     * [VaultPlayerScreen] instead; surfaced here only as a defensive
     * fallback if that dispatch is ever wrong. */
    data class WrongScreen(val message: String) : VaultReaderState()
}

/**
 * Opens one vault file for reading — EPUB via [VaultReadiumProvider], PDF via
 * the [VaultFileReader] directly (the PDF screen builds its own
 * `ParcelFileDescriptor` from it via `core:vaultcontent`'s proxy-fd/memfd
 * adapters). Reading progress is still not persisted (no progress field on
 * [xyz.libravault.core.vaultstore.VaultManifestEntry] yet) but bookmarks and
 * (EPUB-only, matching `feature:reader`'s own PDF gap) highlights are —
 * stored via [VaultStore.addBookmark]/[VaultStore.addHighlight] in the
 * encrypted manifest, same as the rest of a vault entry's metadata.
 *
 * [settings] ([VaultReaderSettings]) is per-user reading preferences —
 * theme/font size/font family/line spacing — same session-only behavior as
 * `feature:reader`'s `ReaderSettings`: nothing here is persisted, it resets
 * to defaults every time the reader screen opens.
 */
@HiltViewModel
class VaultReaderViewModel @Inject constructor(
    private val sessionManager: VaultSessionManager,
    private val readiumProvider: VaultReadiumProvider,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    val vaultId: String = checkNotNull(savedStateHandle["vaultId"]) { "VaultReaderScreen requires a vaultId nav argument" }
    private val fileIdHex: String = checkNotNull(savedStateHandle["fileId"]) { "VaultReaderScreen requires a fileId nav argument" }
    private val fileId: ByteArray = fileIdHex.hexToFileId()

    private val _state = MutableStateFlow<VaultReaderState>(VaultReaderState.Loading)
    val state: StateFlow<VaultReaderState> = _state.asStateFlow()

    private val _settings = MutableStateFlow(VaultReaderSettings())
    val settings: StateFlow<VaultReaderSettings> = _settings.asStateFlow()

    private val _bookmarks = MutableStateFlow<List<VaultBookmark>>(emptyList())
    val bookmarks: StateFlow<List<VaultBookmark>> = _bookmarks.asStateFlow()

    private val _highlights = MutableStateFlow<List<VaultHighlight>>(emptyList())
    val highlights: StateFlow<List<VaultHighlight>> = _highlights.asStateFlow()

    /** The reader's last-reported position — CFI/Locator JSON for EPUB,
     * `"page:N"` for PDF, same convention [VaultBookmark.positionRef] and
     * `core.domain.model.Bookmark.positionRef` both already use. Fed by
     * [onEpubPositionChanged]/[onPdfPageChanged]; [addBookmark] reads it. */
    private val _currentPositionRef = MutableStateFlow<String?>(null)
    val currentPositionRef: StateFlow<String?> = _currentPositionRef.asStateFlow()

    /** Set by [navigateToBookmark] when the user taps a bookmark in the
     * sheet; the screen forwards it to whichever renderer is active and
     * clears it via [clearPendingNavigation] once consumed. */
    private val _pendingNavigationRef = MutableStateFlow<String?>(null)
    val pendingNavigationRef: StateFlow<String?> = _pendingNavigationRef.asStateFlow()

    private var reader: VaultFileReader? = null
    private var publication: Publication? = null
    private var store: VaultStore? = null

    /** Only valid once [state] is [VaultReaderState.PdfReady]. */
    fun pdfReader(): VaultFileReader = requireNotNull(reader) { "PDF reader requested before it was opened" }

    init {
        viewModelScope.launch {
            if (!sessionManager.isUnlocked(vaultId)) {
                _state.value = VaultReaderState.Error("Vault is locked")
                return@launch
            }
            val s = sessionManager.requireUnlocked(vaultId)
            store = s
            val entry = s.listEntries().find { it.fileId.contentEquals(fileId) }
            if (entry == null) {
                _state.value = VaultReaderState.Error("File not found in this vault")
                return@launch
            }
            if (entry.format in VAULT_AUDIO_FORMAT_NAMES) {
                _state.value = VaultReaderState.WrongScreen("This is an audio file — open it from the player instead")
                return@launch
            }
            _bookmarks.value = entry.bookmarks
            _highlights.value = entry.highlights

            val r = s.openReader(fileId)
            reader = r
            when (entry.format) {
                "EPUB" -> readiumProvider.open(r, fileIdHex).fold(
                    onSuccess = { pub -> publication = pub; _state.value = VaultReaderState.EpubReady(entry.title, pub) },
                    onFailure = { e ->
                        _state.value = if (e is VaultDrmProtectedException) {
                            VaultReaderState.DrmProtected(e.schemeName)
                        } else {
                            VaultReaderState.Error(e.message ?: "Could not open EPUB")
                        }
                    },
                )
                "PDF" -> _state.value = VaultReaderState.PdfReady(entry.title)
                else -> _state.value = VaultReaderState.Error("Unsupported format: ${entry.format}")
            }
        }
    }

    override fun onCleared() {
        publication?.close()
        reader?.close()
    }

    // ── Reading settings ──────────────────────────────────────────────────────

    fun onThemeChanged(theme: ReadingTheme) {
        _settings.update { it.copy(theme = theme) }
    }

    fun onFontSizeChanged(size: Float) {
        _settings.update { it.copy(fontSize = size.coerceIn(0.8f, 2.0f)) }
    }

    fun onFontFamilyChanged(family: VaultReaderFontFamily) {
        _settings.update {
            it.copy(
                fontFamily = family,
                // OpenDyslexic bundles a sensible line-spacing default with the
                // font itself (#423) — dyslexia-friendly typography guidance
                // recommends both together. Any other family leaves the user's
                // current line spacing untouched.
                lineSpacing = if (family == VaultReaderFontFamily.OPEN_DYSLEXIC) {
                    VAULT_DYSLEXIA_FRIENDLY_LINE_SPACING
                } else {
                    it.lineSpacing
                },
            )
        }
    }

    fun onLineSpacingChanged(spacing: Float) {
        _settings.update { it.copy(lineSpacing = spacing.coerceIn(1.0f, 2.5f)) }
    }

    fun onWarmthChanged(warmth: Float) {
        _settings.update { it.copy(warmth = warmth.coerceIn(0f, 1f)) }
    }

    fun onScrollModeChanged(mode: VaultScrollMode) {
        _settings.update { it.copy(scrollMode = mode) }
    }

    // ── Position tracking ────────────────────────────────────────────────────

    fun onEpubPositionChanged(locatorJson: String) {
        _currentPositionRef.value = locatorJson
    }

    fun onPdfPageChanged(pageIndex: Int) {
        _currentPositionRef.value = "page:$pageIndex"
    }

    // ── Bookmarks ─────────────────────────────────────────────────────────────

    /** Bookmarks the reader's current position — a no-op if no position has
     * been reported yet (e.g. tapped before the EPUB navigator/PDF renderer
     * finished its first layout pass). */
    fun addBookmark(label: String? = null) {
        val ref = _currentPositionRef.value ?: return
        val s = store ?: return
        viewModelScope.launch {
            val bookmark = s.addBookmark(fileId, ref, label)
            _bookmarks.update { it + bookmark }
        }
    }

    fun removeBookmark(id: Long) {
        val s = store ?: return
        viewModelScope.launch {
            s.removeBookmark(fileId, id)
            _bookmarks.update { list -> list.filterNot { it.id == id } }
        }
    }

    fun updateBookmarkNote(id: Long, note: String?) {
        val s = store ?: return
        viewModelScope.launch {
            s.updateBookmarkNote(fileId, id, note)
            _bookmarks.update { list -> list.map { if (it.id == id) it.copy(note = note) else it } }
        }
    }

    /** Requests navigation to [positionRef] — the screen picks up
     * [pendingNavigationRef] and forwards it to the active renderer. */
    fun navigateToBookmark(positionRef: String) {
        _pendingNavigationRef.value = positionRef
    }

    fun clearPendingNavigation() {
        _pendingNavigationRef.value = null
    }

    // ── Highlights (EPUB only — matches feature:reader's own PDF gap) ─────────

    fun addHighlight(positionRef: String, text: String, colorHex: String = "#FFE066") {
        val s = store ?: return
        viewModelScope.launch {
            val highlight = s.addHighlight(fileId, positionRef, text, colorHex)
            _highlights.update { it + highlight }
        }
    }

    fun removeHighlight(id: Long) {
        val s = store ?: return
        viewModelScope.launch {
            s.removeHighlight(fileId, id)
            _highlights.update { list -> list.filterNot { it.id == id } }
        }
    }
}

File: ios/LibraVaultApp/LibraVault/LibraVault/Sources/Features/EncryptedVault/VaultReaderViewModel.swift
import Foundation
import PDFKit

/// Formats `VaultReaderViewModel`/`VaultPlayerViewModel` route to, matching
/// `VaultManifestEntry.format`'s stored value — always a lowercase
/// `MediaFormat` case name (`VaultStore.importFile`'s caller passes
/// `String(describing: format)`; see `EncryptedVaultContentsViewModel
/// .importOne`).
enum VaultContentFormat {
    static let audioFormats: Set<String> = ["mp3", "m4b", "aac", "flac", "ogg", "opus"]

    static func isAudio(_ format: String) -> Bool { audioFormats.contains(format) }
}

enum VaultReaderState: Equatable {
    case loading
    case error(String)
    case epubReady(title: String)
    case pdfReady(title: String)
    /// `text` is the whole decoded document — v1 renders it as a single
    /// scrollable document via `MarkdownReaderContent`, no TOC/mermaid/
    /// relative-image parity with the non-vault reader yet (issue #442).
    case markdownReady(title: String, text: String)
    /// Not a reading format — the caller should have routed audio to
    /// `VaultPlayerView` instead. Surfaced only as a defensive fallback if
    /// that dispatch is ever wrong, mirroring Android's identical
    /// `VaultReaderState.WrongScreen`.
    case wrongScreen(String)

    static func == (lhs: VaultReaderState, rhs: VaultReaderState) -> Bool {
        switch (lhs, rhs) {
        case (.loading, .loading): return true
        case (.error(let a), .error(let b)): return a == b
        case (.epubReady(let a), .epubReady(let b)): return a == b
        case (.pdfReady(let a), .pdfReady(let b)): return a == b
        case (.markdownReady(let at, let ax), .markdownReady(let bt, let bx)): return at == bt && ax == bx
        case (.wrongScreen(let a), .wrongScreen(let b)): return a == b
        default: return false
        }
    }
}

/// Opens one vault file for reading — EPUB or PDF, decrypted fully into
/// memory via `VaultStore.readFullContent` (never written to a plaintext
/// temp file) and handed to `VaultEPUBParser`/`PDFDocument(data:)`
/// respectively. Bookmarks and highlights persist through
/// `VaultStore.addBookmark`/`addHighlight` in the encrypted manifest — called
/// directly, the same "no shared interface with the non-vault
/// repositories" pattern `VaultStore` itself already establishes (see its
/// own doc comment).
///
/// Reading position is tracked only coarsely — `"chapter:N"` for EPUB,
/// `"page:N"` for PDF — since iOS's existing reader has no CFI-equivalent
/// locator (it renders `BookChapter` text directly, unlike Android's
/// Readium-backed navigator). Reading settings (theme/font/line-spacing)
/// and reading position are intentionally not persisted here — see #203's
/// acceptance-criteria note: Android doesn't persist those for vault
/// content either, only bookmarks/highlights, so iOS matches that rather
/// than over-delivering.
@MainActor
final class VaultReaderViewModel: ObservableObject {

    let vaultId: String
    let fileId: Data

    @Published private(set) var state: VaultReaderState = .loading
    @Published private(set) var chapters: [BookChapter] = []
    @Published private(set) var pdfDocument: PDFDocument?
    @Published private(set) var markdownBlocks: [MarkdownBlock] = []
    @Published private(set) var bookmarks: [VaultBookmark] = []
    @Published private(set) var highlights: [VaultHighlight] = []
    @Published var currentChapterIndex: Int = 0
    @Published var currentPageIndex: Int = 0
    @Published var errorMessage: String?
    /// Flips when the vault locks while this screen is open — either
    /// noticed passively via `checkStillUnlocked()` (#526) or triggered
    /// actively via `lock()` (#668). `VaultReaderView` observes this and
    /// pops back to the unlock flow; nothing here does that itself, mirroring
    /// `EncryptedVaultContentsViewModel.isLocked`'s split between model and
    /// view.
    @Published private(set) var wasLocked = false

    private let sessionManager: VaultSessionManager
    private var store: VaultStore?

    init(vaultId: String, fileId: Data, sessionManager: VaultSessionManager) {
        self.vaultId = vaultId
        self.fileId = fileId
        self.sessionManager = sessionManager
    }

    func load() async {
        guard await sessionManager.isUnlocked(vaultId) else {
            state = .error("Vault is locked")
            return
        }
        let s = await sessionManager.requireUnlocked(vaultId)
        store = s

        do {
            guard let entry = try s.listEntries().first(where: { $0.fileId == fileId }) else {
                state = .error("File not found in this vault")
                return
            }
            bookmarks = entry.bookmarks
            highlights = entry.highlights

            if VaultContentFormat.isAudio(entry.format) {
                state = .wrongScreen("This is an audio file — open it from the player instead")
                return
            }

            let content = try s.readFullContent(fileId: fileId)
            switch entry.format {
            case "epub":
                chapters = try VaultEPUBParser.parse(data: content)
                state = .epubReady(title: entry.title)
            case "pdf":
                guard let document = PDFDocument(data: content) else {
                    state = .error("Could not open PDF")
                    return
                }
                guard document.pageCount > 0 else {
                    state = .error("PDF has no pages")
                    return
                }
                pdfDocument = document
                state = .pdfReady(title: entry.title)
            case "markdown":
                let text = String(decoding: content, as: UTF8.self)
                markdownBlocks = MarkdownDocumentParser.parse(text)
                state = .markdownReady(title: entry.title, text: text)
            default:
                state = .error("Unsupported format: \(entry.format)")
            }
        } catch {
            state = .error(error.localizedDescription)
        }
    }

    // MARK: - Vault lock observation (#526/#668)

    /// Re-checks vault lock state when this screen returns to the
    /// foreground — called from `VaultReaderView`'s `scenePhase` observer,
    /// the SwiftUI equivalent of the `DisposableEffect`+`ON_RESUME` idiom
    /// the unified Android `ReaderScreen` uses for the identical gap (ported
    /// from the deleted `VaultReaderScreen`, see that Kotlin file's own
    /// `checkStillUnlocked()` doc comment). Nothing else here observes
    /// `VaultSessionManager` continuously, so a lock that fired while this
    /// screen was backgrounded (`VaultForegroundLockObserver`) would
    /// otherwise go unnoticed until the user tried and failed to read/mutate
    /// something. No-op while still loading — `load()` itself already
    /// handles the locked case for the initial open.
    func checkStillUnlocked() async {
        guard state != .loading else { return }
        if await !sessionManager.isUnlocked(vaultId) {
            wasLocked = true
        }
    }

    /// Actively locks the vault and notices it via the same `wasLocked` path
    /// `checkStillUnlocked()` uses — called when `.vaultContentSecurity()`
    /// detects a screen recording/AirPlay mirror or an imminent app-switcher
    /// snapshot while vault content is on screen (#668), not just reacting
    /// passively to a lock that already happened elsewhere.
    func lock() async {
        await sessionManager.lock(vaultId)
        wasLocked = true
    }

    // MARK: - Position tracking

    private var currentPositionRef: String {
        switch state {
        case .pdfReady: return "page:\(currentPageIndex)"
        default: return "chapter:\(currentChapterIndex)"
        }
    }

    // MARK: - Bookmarks

    /// Bookmarks the reader's current position. `async` rather than an
    /// internal fire-and-forget `Task` — mirrors
    /// `EncryptedVaultContentsViewModel.lock()`'s convention of pushing the
    /// `Task { await ... }` wrapping out to the caller (a button action),
    /// which is also what lets a test simply `await` the result instead of
    /// polling for it.
    func addBookmark(label: String? = nil) async {
        guard let store else { return }
        let ref = currentPositionRef
        do {
            let bookmark = try store.addBookmark(fileId: fileId, positionRef: ref, label: label)
            bookmarks.append(bookmark)
        } catch VaultStoreError.vaultLocked {
            wasLocked = true
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func removeBookmark(id: Int64) async {
        guard let store else { return }
        do {
            try store.removeBookmark(fileId: fileId, bookmarkId: id)
            bookmarks.removeAll { $0.id == id }
        } catch VaultStoreError.vaultLocked {
            wasLocked = true
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updateBookmarkNote(id: Int64, note: String?) async {
        guard let store else { return }
        do {
            try store.updateBookmarkNote(fileId: fileId, bookmarkId: id, note: note)
            if let index = bookmarks.firstIndex(where: { $0.id == id }) {
                bookmarks[index].note = note
            }
        } catch VaultStoreError.vaultLocked {
            wasLocked = true
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    /// Jumps the active renderer to `bookmark`'s saved position.
    func navigate(to bookmark: VaultBookmark) {
        if bookmark.positionRef.hasPrefix("page:"),
           let page = Int(bookmark.positionRef.dropFirst("page:".count)) {
            currentPageIndex = page
        } else if bookmark.positionRef.hasPrefix("chapter:"),
                  let chapter = Int(bookmark.positionRef.dropFirst("chapter:".count)) {
            currentChapterIndex = chapter
        }
    }

    // MARK: - Highlights

    /// Highlights the currently displayed text — the whole current chapter
    /// for EPUB, or the whole current page's extracted text for PDF. Coarser
    /// than a real text-selection range (iOS's existing reader has no
    /// selection-to-highlight UI at all yet — see `HighlightsSheet`'s own
    /// doc comment), but still round-trips through the encrypted manifest
    /// exactly like a finer-grained highlight would, which is what #203's
    /// acceptance criteria require.
    func addHighlight(colorHex: String = "#FFE066") async {
        guard let store else { return }
        let (ref, text): (String, String)
        switch state {
        case .pdfReady:
            guard let page = pdfDocument?.page(at: currentPageIndex) else { return }
            ref = "page:\(currentPageIndex)"
            text = page.string ?? ""
        default:
            guard currentChapterIndex < chapters.count else { return }
            ref = "chapter:\(currentChapterIndex)"
            text = chapters[currentChapterIndex].text
        }
        guard !text.isEmpty else { return }

        do {
            let highlight = try store.addHighlight(fileId: fileId, positionRef: ref, highlightedText: text, colorHex: colorHex)
            highlights.append(highlight)
        } catch VaultStoreError.vaultLocked {
            wasLocked = true
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func removeHighlight(id: Int64) async {
        guard let store else { return }
        do {
            try store.removeHighlight(fileId: fileId, highlightId: id)
            highlights.removeAll { $0.id == id }
        } catch VaultStoreError.vaultLocked {
            wasLocked = true
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}
